# SQL Basics

## Introduction

This notebook is a short introduction to SQL. It is based on the [DuckDB](https://duckdb.org/) database engine.

## Datasets

The following datasets are used in this notebook. You don't need to download them, they can be accessed directly from the notebook.

- [cities.csv](https://storage.googleapis.com/qm2/CASA0025/cities.csv)
- [countries.csv](https://storage.googleapis.com/qm2/CASA0025/countries.csv)

## References

- [W3Schools SQL Tutorial](https://www.w3schools.com/sql)
- [DuckDB SQL Introduction](https://duckdb.org/docs/sql/introduction.html)

## Installation

Uncomment the following cell to install the required packages.

In [1]:
%pip install duckdb duckdb-engine jupysql

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 49.7/49.7 kB 1.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 95.1/95.1 kB 4.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 192.8/192.8 kB 8.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 168.4/168.4 kB 7.4 MB/s eta 0:00:00


## Library Import and Configuration

In [2]:
import duckdb
import pandas as pd

# Import jupysql Jupyter extension to create SQL cells
%load_ext sql

Set configurations on jupysql to directly output data to Pandas and to simplify the output that is printed to the notebook.

In [3]:
%config SqlMagic.autopandas = True
%config SqlMagic.feedback = False
%config SqlMagic.displaycon = False

## Connecting to DuckDB

Connect jupysql to DuckDB using a SQLAlchemy-style connection string. You may either connect to an in memory DuckDB, or a file backed db.

In [4]:
%sql duckdb:///:memory:
# %sql duckdb:///path/to/file.db

 If your SQL query is one line only, you may use the `%sql` magic command. For multi-line SQL query, you may use the `%%sql` magic command.

## Install extensions


Check available DuckDB extensions.

In [20]:
%%sql

SELECT * FROM duckdb_extensions();

extension_name  loaded  installed  \
0       autocomplete   False      False   
1                aws   False      False   
2              azure   False      False   
3     core_functions    True       True   
4              delta   False      False   
5           ducklake   False      False   
6          encodings   False      False   
7              excel   False      False   
8                fts   False      False   
9             httpfs    True       True   
10           iceberg   False      False   
11               icu    True       True   
12              inet   False      False   
13          jemalloc    True       True   
14              json    True       True   
15        motherduck   False      False   
16     mysql_scanner   False      False   
17           parquet    True       True   
18  postgres_scanner   False      False   
19           spatial   False      False   
20    sqlite_scanner   False      False   
21             tpcds   False      False   
22              tpch   False      False   
23                ui   False      False   
24               vss   False      False   

                                         install_path  \
0                                                       
1                                                       
2                                                       
3                                          (BUILT-IN)   
4                                                       
5                                                       
6                                                       
7                                                       
8                                                       
9   /root/.duckdb/extensions/v1.3.2/linux_amd64/ht...   
10                                                      
11                                         (BUILT-IN)   
12                                                      
13                                         (BUILT-IN)   
14                                         (BUILT-IN)   
15                                                      
16                                                      
17                                         (BUILT-IN)   
18                                                      
19                                                      
20                                                      
21                                                      
22                                                      
23                                                      
24                                                      

                                          description            aliases  \
0          Adds support for autocomplete in the shell                 []   
1        Provides features that depend on the AWS SDK                 []   
2   Adds a filesystem abstraction for Azure blob s...                 []   
3                               Core function library                 []   
4                         Adds support for Delta Lake                 []   
5   Adds support for DuckLake, SQL as a Lakehouse ...                 []   
6                      All unicode encodings to UTF-8                 []   
7          Adds support for Excel-like format strings                 []   
8           Adds support for Full-Text Search Indexes                 []   
9   Adds support for reading and writing files ove...  [http, https, s3]   
10                    Adds support for Apache Iceberg                 []   
11  Adds support for time zones and collations usi...                 []   
12  Adds support for IP-related data types and fun...                 []   
13          Overwrites system allocator with JEMalloc                 []   
14                   Adds support for JSON operations                 []   
15     Enables motherduck integration with the system               [md]   
16    Adds support for connecting to a MySQL database            [mysql]   
17  Adds support for reading and wri

DuckDB's [httpfs extension](https://duckdb.org/docs/extensions/httpfs) allows parquet and csv files to be queried remotely over http. This is useful for querying large datasets without having to download them locally. Let's install the extension and load the extension.

In [21]:
%%sql

INSTALL httpfs;
LOAD httpfs;

Empty DataFrame
Columns: [Success]
Index: []

## Read CSV

Use the `httpfs` extension to read the `cities.csv` file from the web.

In [7]:
%%sql

SELECT * FROM 'https://storage.googleapis.com/qm2/CASA0025/cities.csv';

id            name country  latitude  longitude  population
0        1           Bombo     UGA   0.58330   32.53330       75000
1        2     Fort Portal     UGA   0.67100   30.27500       42670
2        3         Potenza     ITA  40.64200   15.79900       69060
3        4      Campobasso     ITA  41.56300   14.65600       50762
4        5           Aosta     ITA  45.73700    7.31500       34062
...    ...             ...     ...       ...        ...         ...
1244  1245  Rio de Janeiro     BRA -22.92502  -43.22502    11748000
1245  1246       Sao Paulo     BRA -23.55868  -46.62502    18845000
1246  1247          Sydney     AUS -33.92001  151.18518     4630000
1247  1248       Singapore     SGP   1.29303  103.85582     5183700
1248  1249       Hong Kong     CHN  22.30498  114.18501     7206000

[1249 rows x 6 columns]

In [8]:
%%sql

SELECT * FROM 'https://storage.googleapis.com/qm2/CASA0025/countries.csv';

id            Country Alpha2_code Alpha3_code  Numeric_code  Latitude  \
0      1        Afghanistan          AF         AFG             4   33.0000   
1      2            Albania          AL         ALB             8   41.0000   
2      3            Algeria          DZ         DZA            12   28.0000   
3      4     American Samoa          AS         ASM            16  -14.3333   
4      5            Andorra          AD         AND            20   42.5000   
..   ...                ...         ...         ...           ...       ...   
238  239  Wallis and Futuna          WF         WLF           876  -13.3000   
239  240     Western Sahara          EH         ESH           732   24.5000   
240  241              Yemen          YE         YEM           887   15.0000   
241  242             Zambia          ZM         ZMB           894  -15.0000   
242  243           Zimbabwe          ZW         ZWE           716  -20.0000   

     Longitude  
0         65.0  
1         20.0  
2          3.0  
3       -170.0  
4          1.6  
..         ...  
238     -176.2  
239      -13.0  
240       48.0  
241       30.0  
242       30.0  

[243 rows x 7 columns]

## Create Table

Create a table named `cities` from the `cities.csv` file.

In [9]:
%%sql

CREATE TABLE cities AS SELECT * FROM 'https://storage.googleapis.com/qm2/CASA0025/cities.csv';

Empty DataFrame
Columns: [Success]
Index: []

Create a table named `countries` from the `countries.csv` file.

In [10]:
%%sql

CREATE TABLE countries AS SELECT * FROM 'https://storage.googleapis.com/qm2/CASA0025/countries.csv';

Empty DataFrame
Columns: [Success]
Index: []

Display the table content in the database.

In [11]:
%%sql

FROM cities;

id            name country  latitude  longitude  population
0        1           Bombo     UGA   0.58330   32.53330       75000
1        2     Fort Portal     UGA   0.67100   30.27500       42670
2        3         Potenza     ITA  40.64200   15.79900       69060
3        4      Campobasso     ITA  41.56300   14.65600       50762
4        5           Aosta     ITA  45.73700    7.31500       34062
...    ...             ...     ...       ...        ...         ...
1244  1245  Rio de Janeiro     BRA -22.92502  -43.22502    11748000
1245  1246       Sao Paulo     BRA -23.55868  -46.62502    18845000
1246  1247          Sydney     AUS -33.92001  151.18518     4630000
1247  1248       Singapore     SGP   1.29303  103.85582     5183700
1248  1249       Hong Kong     CHN  22.30498  114.18501     7206000

[1249 rows x 6 columns]

In [12]:
%%sql

FROM countries;

id            Country Alpha2_code Alpha3_code  Numeric_code  Latitude  \
0      1        Afghanistan          AF         AFG             4   33.0000   
1      2            Albania          AL         ALB             8   41.0000   
2      3            Algeria          DZ         DZA            12   28.0000   
3      4     American Samoa          AS         ASM            16  -14.3333   
4      5            Andorra          AD         AND            20   42.5000   
..   ...                ...         ...         ...           ...       ...   
238  239  Wallis and Futuna          WF         WLF           876  -13.3000   
239  240     Western Sahara          EH         ESH           732   24.5000   
240  241              Yemen          YE         YEM           887   15.0000   
241  242             Zambia          ZM         ZMB           894  -15.0000   
242  243           Zimbabwe          ZW         ZWE           716  -20.0000   

     Longitude  
0         65.0  
1         20.0  
2          3.0  
3       -170.0  
4          1.6  
..         ...  
238     -176.2  
239      -13.0  
240       48.0  
241       30.0  
242       30.0  

[243 rows x 7 columns]

## The SQL SELECT statement

The `SELECT` statement is used to select data from a database. Use either `SELECT *` to select all columns, or `SELECT column1, column2, ...` to select specific columns.

`SELECT * FROM cities` is the same as `FROM cities`.

In [13]:
%%sql # select * means everything

SELECT * FROM cities;

id            name country  latitude  longitude  population
0        1           Bombo     UGA   0.58330   32.53330       75000
1        2     Fort Portal     UGA   0.67100   30.27500       42670
2        3         Potenza     ITA  40.64200   15.79900       69060
3        4      Campobasso     ITA  41.56300   14.65600       50762
4        5           Aosta     ITA  45.73700    7.31500       34062
...    ...             ...     ...       ...        ...         ...
1244  1245  Rio de Janeiro     BRA -22.92502  -43.22502    11748000
1245  1246       Sao Paulo     BRA -23.55868  -46.62502    18845000
1246  1247          Sydney     AUS -33.92001  151.18518     4630000
1247  1248       Singapore     SGP   1.29303  103.85582     5183700
1248  1249       Hong Kong     CHN  22.30498  114.18501     7206000

[1249 rows x 6 columns]

To limit the number of rows returned, use the `LIMIT` keyword. For example, `SELECT * FROM cities LIMIT 10` will return only the first 10 rows.

In [14]:
%%sql

SELECT * FROM cities LIMIT 10;

id              name country  latitude  longitude  population
0   1             Bombo     UGA   0.58330   32.53330       75000
1   2       Fort Portal     UGA   0.67100   30.27500       42670
2   3           Potenza     ITA  40.64200   15.79900       69060
3   4        Campobasso     ITA  41.56300   14.65600       50762
4   5             Aosta     ITA  45.73700    7.31500       34062
5   6         Mariehamn     ALD  60.09700   19.94900       10682
6   7          Ramallah     PSE  31.90294   35.20621       24599
7   8      Vatican City     VAT  41.90001   12.44781         832
8   9           Poitier     FRA  46.58329    0.33328       85960
9  10  Clermont-Ferrand     FRA  45.77998    3.08001      233050

Select a subset of columns from the `cities` table and display the first 10 rows.

In [15]:
%%sql

SELECT name, country FROM cities LIMIT 10;

name country
0             Bombo     UGA
1       Fort Portal     UGA
2           Potenza     ITA
3        Campobasso     ITA
4             Aosta     ITA
5         Mariehamn     ALD
6          Ramallah     PSE
7      Vatican City     VAT
8           Poitier     FRA
9  Clermont-Ferrand     FRA

To select distinct values, use the `DISTINCT` keyword. For example, `SELECT DISTINCT country FROM cities` will return only the distinct values of the `country` column.

In [16]:
%%sql

SELECT DISTINCT country FROM cities LIMIT 10;

country
0     UGA
1     ITA
2     PSE
3     VAT
4     NAM
5     NLD
6     LIE
7     GEO
8     PRT
9     MHL

To count the number of rows returned, use the `COUNT(*)` function. For example, `SELECT COUNT(*) FROM cities` will return the number of rows in the `cities` table.

In [17]:
%%sql

SELECT COUNT(*) FROM cities;

count_star()
0          1249

To count the number of distinct values, use the `COUNT(DISTINCT column)` function. For example, `SELECT COUNT(DISTINCT country) FROM cities` will return the number of distinct values in the `country` column.

In [18]:
%%sql

SELECT COUNT(DISTINCT country) FROM cities;

count(DISTINCT country)
0                      200

To calculate the maximum value, use the `MAX(column)` function. For example, `SELECT MAX(population) FROM cities` will return the maximum value in the `population` column.

In [22]:
%%sql

SELECT MAX(population) FROM cities;

max(population)
0         35676000

To calculate the total value, use the `SUM(column)` function. For example, `SELECT SUM(population) FROM cities` will return the total value in the `population` column.

In [23]:
%%sql

SELECT SUM(population) FROM cities;

sum(population)
0     1.475535e+09

To calculate the average value, use the `AVG(column)` function. For example, `SELECT AVG(population) FROM cities` will return the average value in the `population` column.

In [24]:
%%sql

SELECT AVG(population) FROM cities;

avg(population)
0     1.181373e+06

To order the results, use the `ORDER BY column` clause. For example, `SELECT * FROM cities ORDER BY country` will return the rows ordered by the `country` column alphabetically.

In [25]:
%%sql

SELECT * FROM cities ORDER BY country LIMIT 10;

id            name country  latitude  longitude  population
0   665           Herat     AFG  34.33001   62.16999      481009
1   666  Mazar-e Sharif     AFG  36.69999   67.10003      458151
2  1024        Kandahar     AFG  31.61002   65.69495      715542
3  1219           Kabul     AFG  34.51669   69.18326     3277000
4   669         Malanje     AGO  -9.54000   16.34003      125856
5   670        Benguela     AGO -12.57826   13.40723      151226
6   671         Lubango     AGO -14.91001   13.49002      125632
7  1026        Menongue     AGO -14.66661   17.69999       13030
8  1027          Huambo     AGO -12.74999   15.76001     1100000
9  1177          Luanda     AGO  -8.83829   13.23443     5172900

To order the results in descending order, use the `ORDER BY column DESC` clause. For example, `SELECT * FROM cities ORDER BY country ASC, population DESC` will return the rows ordered by the `country` column alphabetical order and then by the `population` column in descending order.

In [26]:
%%sql

SELECT * FROM cities ORDER BY country ASC, population DESC LIMIT 10;

id            name country  latitude  longitude  population
0  1219           Kabul     AFG  34.51669   69.18326     3277000
1  1024        Kandahar     AFG  31.61002   65.69495      715542
2   665           Herat     AFG  34.33001   62.16999      481009
3   666  Mazar-e Sharif     AFG  36.69999   67.10003      458151
4  1177          Luanda     AGO  -8.83829   13.23443     5172900
5  1027          Huambo     AGO -12.74999   15.76001     1100000
6   670        Benguela     AGO -12.57826   13.40723      151226
7   672          Namibe     AGO -15.19004   12.16002      132900
8   669         Malanje     AGO  -9.54000   16.34003      125856
9   671         Lubango     AGO -14.91001   13.49002      125632

## The WHERE Clause

The `WHERE` clause is used to filter records. The `WHERE` clause is used to extract only those records that fulfill a specified condition.

In [19]:
%%sql

SELECT * FROM cities WHERE country='USA'

id                 name country  latitude  longitude  population
0     115                Agana     USA  13.47002  144.75002      122411
1     124       San Bernardino     USA  34.12038 -117.30003     1745000
2     125           Bridgeport     USA  41.17998  -73.19996     1018000
3     126            Rochester     USA  43.17043  -77.61995      755000
4     160  International Falls     USA  48.60113  -93.41085       15240
..    ...                  ...     ...       ...        ...         ...
109  1186              Atlanta     USA  33.83001  -84.39995     4506000
110  1187              Chicago     USA  41.82999  -87.75005     8990000
111  1223          Los Angeles     USA  33.98998 -118.17998    12500000
112  1224      Washington D.C.     USA  38.89955  -77.00942     4338000
113  1225             New York     USA  40.74998  -73.98002    19040000

[114 rows x 6 columns]

You can use boolean operators such as `AND`, `OR`, `NOT` to filter records. For example, `SELECT * FROM cities WHERE country='USA' OR country='CAN'` will return the rows where the `country` column is either `USA` or `CAN`.

In [27]:
%%sql

SELECT * FROM cities WHERE country='USA' OR country='CAN';

id                 name country  latitude  longitude  population
0     115                Agana     USA  13.47002  144.75002      122411
1     124       San Bernardino     USA  34.12038 -117.30003     1745000
2     125           Bridgeport     USA  41.17998  -73.19996     1018000
3     126            Rochester     USA  43.17043  -77.61995      755000
4     160  International Falls     USA  48.60113  -93.41085       15240
..    ...                  ...     ...       ...        ...         ...
154  1216            Vancouver     CAN  49.27342 -123.12164     2313328
155  1217              Toronto     CAN  43.69998  -79.42002     5213000
156  1223          Los Angeles     USA  33.98998 -118.17998    12500000
157  1224      Washington D.C.     USA  38.89955  -77.00942     4338000
158  1225             New York     USA  40.74998  -73.98002    19040000

[159 rows x 6 columns]

To select US cities with a population greater than 1 million, use the following query: `SELECT * FROM cities WHERE country='USA' AND population > 1000000`.

In [28]:
%%sql

SELECT * FROM cities WHERE country='USA' AND population>1000000;

id             name country  latitude  longitude  population
0    124   San Bernardino     USA  34.12038 -117.30003     1745000
1    125       Bridgeport     USA  41.17998  -73.19996     1018000
2    183      Kansas City     USA  39.10709  -94.60409     1469000
3    188        Ft. Worth     USA  32.73998  -97.34004     1440454
4    190           Austin     USA  30.26695  -97.74278     1161000
5    198          Orlando     USA  28.50998  -81.38003     1350000
6    202     Indianapolis     USA  39.74999  -86.17005     1436000
7    207        Baltimore     USA  39.29999  -76.61998     2255000
8    735         San Jose     USA  37.29998 -121.84999     1668000
9    736       Sacramento     USA  38.57502 -121.47004     1604000
10   737        Las Vegas     USA  36.21000 -115.22001     1823000
11   739         Portland     USA  45.52002 -122.67999     1875000
12   746      San Antonio     USA  29.48733  -98.50731     1473000
13   750          Raleigh     USA  35.81878  -78.64469     1163515
14   751        Cleveland     USA  41.46999  -81.69500     1890000
15   752       Cincinnati     USA  39.16188  -84.45692     1636000
16   754          Memphis     USA  35.11999  -90.00000     1081000
17   755          Norfolk     USA  36.84996  -76.28001     1047869
18   756        Milwaukee     USA  43.05266  -87.91997     1388000
19   757          Buffalo     USA  42.87998  -78.88000     1016000
20   758       Pittsburgh     USA  40.43000  -79.99999     1838000
21  1049         San Juan     USA  18.44002  -66.12998     2478905
22  1066      Minneapolis     USA  44.97998  -93.25179     2616000
23  1068          Seattle     USA  47.57000 -122.33998     3074000
24  1069          Phoenix     USA  33.53998 -112.06999     3551000
25  1070        San Diego     USA  32.82002 -117.17999     2916000
26  1071        St. Louis     USA  38.63502  -90.23998     2199000
27  1073           Dallas     USA  32.82002  -96.84002     4798000
28  1075           Boston     USA  42.32996  -71.07001     4467000
29  1076            Tampa     USA  27.94699  -82.45862     2314000
30  1077     Philadelphia     USA  39.99997  -75.17000     5492000
31  1078          Detroit     USA  42.32996  -83.08006     4101000
32  1182    San Francisco     USA  37.74001 -122.45998     3450000
33  1183           Denver     USA  39.73919 -104.98402     2313000
34  1184          Houston     USA  29.81997  -95.33998     4459000
35  1185            Miami     USA  25.78761  -80.22411     5585000
36  1186          Atlanta     USA  33.83001  -84.39995     4506000
37  1187          Chicago     USA  41.82999  -87.75005     8990000
38  1223      Los Angeles     USA  33.98998 -118.17998    12500000
39  1224  Washington D.C.     USA  38.89955  -77.00942     4338000
40  1225         New York     USA  40.74998  -73.98002    19040000

To select cities with the country name starting with the letter `U`, use the following query: `SELECT * FROM cities WHERE country LIKE 'U%'`.

In [29]:
%%sql

SELECT * FROM cities WHERE country LIKE 'U%';

id             name country  latitude  longitude  population
0       1            Bombo     UGA   0.58330   32.53330       75000
1       2      Fort Portal     UGA   0.67100   30.27500       42670
2      14            Jinja     UGA   0.44042   33.19993      301619
3      15             Arua     UGA   3.02037   30.90002      250000
4      16            Mbale     UGA   1.09041   34.17000      402368
..    ...              ...     ...       ...        ...         ...
134  1189             Kiev     UKR  50.43337   30.51663     2709000
135  1191         Tashkent     UZB  41.31170   69.29493     2184000
136  1223      Los Angeles     USA  33.98998 -118.17998    12500000
137  1224  Washington D.C.     USA  38.89955  -77.00942     4338000
138  1225         New York     USA  40.74998  -73.98002    19040000

[139 rows x 6 columns]

To select cities with the country name ending with the letter `A`, use the following query: `SELECT * FROM cities WHERE country LIKE '%A'`.

In [30]:
%%sql

SELECT * FROM cities WHERE country LIKE '%A';

id            name country  latitude  longitude  population
0       1           Bombo     UGA   0.58330   32.53330       75000
1       2     Fort Portal     UGA   0.67100   30.27500       42670
2       3         Potenza     ITA  40.64200   15.79900       69060
3       4      Campobasso     ITA  41.56300   14.65600       50762
4       5           Aosta     ITA  45.73700    7.31500       34062
..    ...             ...     ...       ...        ...         ...
302  1232           Lagos     NGA   6.44326    3.39153     9466000
303  1233            Rome     ITA  41.89596   12.48326     3339000
304  1242           Paris     FRA  48.86669    2.33334     9904000
305  1245  Rio de Janeiro     BRA -22.92502  -43.22502    11748000
306  1246       Sao Paulo     BRA -23.55868  -46.62502    18845000

[307 rows x 6 columns]

To select cities with the country name containing the letter `S` in the middle, use the following query: `SELECT * FROM cities WHERE country LIKE '_S_'`.

In [ ]:
%%sql

SELECT * FROM cities WHERE country LIKE '_S_';

To select cities from a list of countries, use the `IN` operator. For example, `SELECT * FROM cities WHERE country IN ('USA', 'CAN')` will return the rows where the `country` column is either `USA` or `CAN`.

In [ ]:
%%sql

SELECT * FROM cities WHERE country IN ('USA', 'CAN');

To select cities with a population between 1 and 10 million, use the following query: `SELECT * FROM cities WHERE population BETWEEN 1000000 AND 10000000`.

In [31]:
%%sql

SELECT * FROM cities WHERE population BETWEEN 1000000 AND 10000000;

id            name country  latitude  longitude  population
0      98           Turin     ITA  45.07039    7.66996     1652000
1     104           Lille     FRA  50.64997    3.08001     1044000
2     124  San Bernardino     USA  34.12038 -117.30003     1745000
3     125      Bridgeport     USA  41.17998  -73.19996     1018000
4     127      Manchester     GBR  53.50042   -2.24799     2230000
..    ...             ...     ...       ...        ...         ...
371  1242           Paris     FRA  48.86669    2.33334     9904000
372  1243        Santiago     CHL -33.45001  -70.66704     5720000
373  1247          Sydney     AUS -33.92001  151.18518     4630000
374  1248       Singapore     SGP   1.29303  103.85582     5183700
375  1249       Hong Kong     CHN  22.30498  114.18501     7206000

[376 rows x 6 columns]

## SQL Joins

Reference: https://www.w3schools.com/sql/sql_join.asp

Here are the different types of the JOINs in SQL:

- `(INNER) JOIN`: Returns records that have matching values in both tables
- `LEFT (OUTER) JOIN`: Returns all records from the left table, and the matched records from the right table
- `RIGHT (OUTER) JOIN`: Returns all records from the right table, and the matched records from the left table
- `FULL (OUTER) JOIN`: Returns all records when there is a match in either left or right table

![](https://i.imgur.com/mITYzuS.png)

We have two sample tables: `cities` and `countries`.

There are 1,249 cities in the `cities` table and 243 countries in the `countries` table.

In [32]:
%%sql

SELECT COUNT(*) FROM cities;

count_star()
0          1249

In [33]:
%%sql

SELECT * FROM cities LIMIT 10;

id              name country  latitude  longitude  population
0   1             Bombo     UGA   0.58330   32.53330       75000
1   2       Fort Portal     UGA   0.67100   30.27500       42670
2   3           Potenza     ITA  40.64200   15.79900       69060
3   4        Campobasso     ITA  41.56300   14.65600       50762
4   5             Aosta     ITA  45.73700    7.31500       34062
5   6         Mariehamn     ALD  60.09700   19.94900       10682
6   7          Ramallah     PSE  31.90294   35.20621       24599
7   8      Vatican City     VAT  41.90001   12.44781         832
8   9           Poitier     FRA  46.58329    0.33328       85960
9  10  Clermont-Ferrand     FRA  45.77998    3.08001      233050

In [34]:
%%sql

SELECT COUNT(*) FROM countries;

count_star()
0           243

In [35]:
%%sql

SELECT * FROM countries LIMIT 10;

id              Country Alpha2_code Alpha3_code  Numeric_code  Latitude  \
0   1          Afghanistan          AF         AFG             4   33.0000   
1   2              Albania          AL         ALB             8   41.0000   
2   3              Algeria          DZ         DZA            12   28.0000   
3   4       American Samoa          AS         ASM            16  -14.3333   
4   5              Andorra          AD         AND            20   42.5000   
5   6               Angola          AO         AGO            24  -12.5000   
6   7             Anguilla          AI         AIA           660   18.2500   
7   8           Antarctica          AQ         ATA            10  -90.0000   
8   9  Antigua and Barbuda          AG         ATG            28   17.0500   
9  10            Argentina          AR         ARG            32  -34.0000   

   Longitude  
0    65.0000  
1    20.0000  
2     3.0000  
3  -170.0000  
4     1.6000  
5    18.5000  
6   -63.1667  
7     0.0000  
8   -61.8000  
9   -64.0000

### SQL Inner Join

The `INNER JOIN` keyword selects records that have matching values in both tables. In the example, we join the `cities` table with the `countries` table using the `country` column in the `cities` table and the `Alpha3_code` column in the `countries` table. The result contains 1,244 rows, indicating that there are 5 cities that do not have a matching country.

In [36]:
%%sql

SELECT * FROM cities INNER JOIN countries ON cities.country = countries."Alpha3_code";

id            name country  latitude  longitude  population  id_1  \
0        1           Bombo     UGA   0.58330   32.53330       75000   226   
1        2     Fort Portal     UGA   0.67100   30.27500       42670   226   
2        3         Potenza     ITA  40.64200   15.79900       69060   106   
3        4      Campobasso     ITA  41.56300   14.65600       50762   106   
4        5           Aosta     ITA  45.73700    7.31500       34062   106   
...    ...             ...     ...       ...        ...         ...   ...   
1239  1245  Rio de Janeiro     BRA -22.92502  -43.22502    11748000    30   
1240  1246       Sao Paulo     BRA -23.55868  -46.62502    18845000    30   
1241  1247          Sydney     AUS -33.92001  151.18518     4630000    13   
1242  1248       Singapore     SGP   1.29303  103.85582     5183700   195   
1243  1249       Hong Kong     CHN  22.30498  114.18501     7206000    43   

      Country_1 Alpha2_code Alpha3_code  Numeric_code  Latitude_1  Longitude_1  
0        Uganda          UG         UGA           800      1.0000      32.0000  
1        Uganda          UG         UGA           800      1.0000      32.0000  
2         Italy          IT         ITA           380     42.8333      12.8333  
3         Italy          IT         ITA           380     42.8333      12.8333  
4         Italy          IT         ITA           380     42.8333      12.8333  
...         ...         ...         ...           ...         ...          ...  
1239     Brazil          BR         BRA            76    -10.0000     -55.0000  
1240     Brazil          BR         BRA            76    -10.0000     -55.0000  
1241  Australia          AU         AUS            36    -27.0000     133.0000  
1242  Singapore          SG         SGP           702      1.3667     103.8000  
1243      China          CN         CHN           156     35.0000     105.0000  

[1244 rows x 13 columns]

Only select the `city` and `country` columns from the `cities` table and the `Country` column from the `countries` table.

In [37]:
%%sql

SELECT name, cities."country", countries."Country" FROM cities INNER JOIN countries ON cities.country = countries."Alpha3_code";

name country  Country_1
0              Bombo     UGA     Uganda
1        Fort Portal     UGA     Uganda
2            Potenza     ITA      Italy
3         Campobasso     ITA      Italy
4              Aosta     ITA      Italy
...              ...     ...        ...
1239  Rio de Janeiro     BRA     Brazil
1240       Sao Paulo     BRA     Brazil
1241          Sydney     AUS  Australia
1242       Singapore     SGP  Singapore
1243       Hong Kong     CHN      China

[1244 rows x 3 columns]

### SQL Left Join

The `LEFT JOIN` keyword returns all records from the left table (`cities`), and the matched records from the right table (`countries`). The result contains 1,249 rows, the same number of rows as the `cities` table.

In [38]:
%%sql

SELECT * FROM cities LEFT JOIN countries ON cities.country = countries."Alpha3_code";

id         name country  latitude  longitude  population  id_1  \
0        1        Bombo     UGA   0.58330   32.53330       75000   226   
1        2  Fort Portal     UGA   0.67100   30.27500       42670   226   
2        3      Potenza     ITA  40.64200   15.79900       69060   106   
3        4   Campobasso     ITA  41.56300   14.65600       50762   106   
4        5        Aosta     ITA  45.73700    7.31500       34062   106   
...    ...          ...     ...       ...        ...         ...   ...   
1244     6    Mariehamn     ALD  60.09700   19.94900       10682  <NA>   
1245   118   Bir Lehlou     SAH  26.11917   -9.65252         500  <NA>   
1246   415     Pristina     KOS  42.66671   21.16598      465186  <NA>   
1247   719      Berbera     SOL  10.43555   45.01641      242344  <NA>   
1248  1053     Hargeysa     SOL   9.56002   44.06531      477876  <NA>   

     Country_1 Alpha2_code Alpha3_code  Numeric_code  Latitude_1  Longitude_1  
0       Uganda          UG         UGA           800      1.0000      32.0000  
1       Uganda          UG         UGA           800      1.0000      32.0000  
2        Italy          IT         ITA           380     42.8333      12.8333  
3        Italy          IT         ITA           380     42.8333      12.8333  
4        Italy          IT         ITA           380     42.8333      12.8333  
...        ...         ...         ...           ...         ...          ...  
1244      None        None        None          <NA>         NaN          NaN  
1245      None        None        None          <NA>         NaN          NaN  
1246      None        None        None          <NA>         NaN          NaN  
1247      None        None        None          <NA>         NaN          NaN  
1248      None        None        None          <NA>         NaN          NaN  

[1249 rows x 13 columns]

### SQL Right Join

The `RIGHT JOIN` keyword returns all records from the right table (`countries`), and the matched records from the left table (`cities`). The result contains 1,291 rows.

In [39]:
%%sql

SELECT * FROM cities RIGHT JOIN countries ON cities.country = countries."Alpha3_code";

id         name country  latitude  longitude  population  id_1  \
0        1        Bombo     UGA    0.5833    32.5333       75000   226   
1        2  Fort Portal     UGA    0.6710    30.2750       42670   226   
2        3      Potenza     ITA   40.6420    15.7990       69060   106   
3        4   Campobasso     ITA   41.5630    14.6560       50762   106   
4        5        Aosta     ITA   45.7370     7.3150       34062   106   
...    ...          ...     ...       ...        ...         ...   ...   
1286  <NA>         None    None       NaN        NaN        <NA>   224   
1287  <NA>         None    None       NaN        NaN        <NA>    88   
1288  <NA>         None    None       NaN        NaN        <NA>   136   
1289  <NA>         None    None       NaN        NaN        <NA>   139   
1290  <NA>         None    None       NaN        NaN        <NA>   185   

                      Country_1 Alpha2_code Alpha3_code  Numeric_code  \
0                        Uganda          UG         UGA           800   
1                        Uganda          UG         UGA           800   
2                         Italy          IT         ITA           380   
3                         Italy          IT         ITA           380   
4                         Italy          IT         ITA           380   
...                         ...         ...         ...           ...   
1286   Turks and Caicos Islands          TC         TCA           796   
1287                   Guernsey          GG         GGY           831   
1288                 Martinique          MQ         MTQ           474   
1289                    Mayotte          YT         MYT           175   
1290  Saint Pierre and Miquelon          PM         SPM           666   

      Latitude_1  Longitude_1  
0         1.0000      32.0000  
1         1.0000      32.0000  
2        42.8333      12.8333  
3        42.8333      12.8333  
4        42.8333      12.8333  
...          ...          ...  
1286     21.7500     -71.5833  
1287     49.5000      -2.5600  
1288     14.6667     -61.0000  
1289    -12.8333      45.1667  
1290     46.8333     -56.3333  

[1291 rows x 13 columns]

### SQL Full Join

The `FULL JOIN` keyword returns all records when there is a match in either left or right table. The result contains 1,296 rows.

In [40]:
%%sql

SELECT * FROM cities FULL JOIN countries ON cities.country = countries."Alpha3_code";

id         name country  latitude  longitude  population  id_1  \
0        1        Bombo     UGA    0.5833    32.5333       75000   226   
1        2  Fort Portal     UGA    0.6710    30.2750       42670   226   
2        3      Potenza     ITA   40.6420    15.7990       69060   106   
3        4   Campobasso     ITA   41.5630    14.6560       50762   106   
4        5        Aosta     ITA   45.7370     7.3150       34062   106   
...    ...          ...     ...       ...        ...         ...   ...   
1291  <NA>         None    None       NaN        NaN        <NA>   224   
1292  <NA>         None    None       NaN        NaN        <NA>    88   
1293  <NA>         None    None       NaN        NaN        <NA>   136   
1294  <NA>         None    None       NaN        NaN        <NA>   139   
1295  <NA>         None    None       NaN        NaN        <NA>   185   

                      Country_1 Alpha2_code Alpha3_code  Numeric_code  \
0                        Uganda          UG         UGA           800   
1                        Uganda          UG         UGA           800   
2                         Italy          IT         ITA           380   
3                         Italy          IT         ITA           380   
4                         Italy          IT         ITA           380   
...                         ...         ...         ...           ...   
1291   Turks and Caicos Islands          TC         TCA           796   
1292                   Guernsey          GG         GGY           831   
1293                 Martinique          MQ         MTQ           474   
1294                    Mayotte          YT         MYT           175   
1295  Saint Pierre and Miquelon          PM         SPM           666   

      Latitude_1  Longitude_1  
0         1.0000      32.0000  
1         1.0000      32.0000  
2        42.8333      12.8333  
3        42.8333      12.8333  
4        42.8333      12.8333  
...          ...          ...  
1291     21.7500     -71.5833  
1292     49.5000      -2.5600  
1293     14.6667     -61.0000  
1294    -12.8333      45.1667  
1295     46.8333     -56.3333  

[1296 rows x 13 columns]

### SQL Union

The `UNION` operator is used to combine the result-set of two or more `SELECT` statements.

In [41]:
%%sql

SELECT country FROM cities
UNION
SELECT "Alpha3_code" FROM countries;

country
0       UGA
1       ALD
2       VAT
3       FRA
4       ATA
..      ...
242     SJM
243     UMI
244     VGB
245     WLF
246     ESH

[247 rows x 1 columns]

## Aggregation

### Group By

The `GROUP BY` statement groups rows that have the same values into summary rows, like "find the number of cities in each country".

The `GROUP BY` statement is often used with aggregate functions (`COUNT`, `MAX`, `MIN`, `SUM`, `AVG`) to group the result-set by one or more columns.

In [ ]:
%%sql

SELECT COUNT(name), country
FROM cities
GROUP BY country
ORDER BY COUNT(name) DESC;

In [ ]:
%%sql

SELECT countries."Country", COUNT(name)
FROM cities
LEFT JOIN countries ON cities.country = countries."Alpha3_code"
GROUP BY countries."Country"
ORDER BY COUNT(name) DESC;

### Having

The `HAVING` clause was added to SQL because the `WHERE` keyword could not be used with aggregate functions.

For example, to select countries with more than 40 cities:

In [ ]:
%%sql

SELECT COUNT(name), country
FROM cities
GROUP BY country
HAVING COUNT(name) > 40
ORDER BY COUNT(name) DESC;

In [ ]:
%%sql

SELECT countries."Country", COUNT(name)
FROM cities
LEFT JOIN countries ON cities.country = countries."Alpha3_code"
GROUP BY countries."Country"
HAVING COUNT(name) > 40
ORDER BY COUNT(name) DESC;

## Conditional statements

The `CASE` statement goes through conditions and returns a value when the first condition is met (like an `IF-THEN-ELSE` statement). So, once a condition is true, it will stop reading and return the result. If no conditions are true, it returns the value in the `ELSE` clause.

For example, to divide cities into 3 groups based on their population:

In [ ]:
%%sql

SELECT name, population,
CASE
    WHEN population > 10000000 THEN 'Megacity'
    WHEN population > 1000000 THEN 'Large city'
    ELSE 'Small city'
END AS category
FROM cities;

## Saving results

You can save the results of a query to a new table using the `CREATE TABLE AS` statement.

In [ ]:
%%sql

CREATE TABLE cities2 AS SELECT * FROM cities;

Show the new table content.

In [ ]:
%%sql

FROM cities2;

Use the `DROP TABLE` statement to delete the table.

In [ ]:
%%sql

DROP TABLE IF EXISTS cities_usa;
CREATE TABLE cities_usa AS (SELECT * FROM cities WHERE country = 'USA');

In [ ]:
%%sql

FROM cities_usa;

Use the `INSERT INTO` statement to insert rows into a table.

In [ ]:
%%sql

INSERT INTO cities_usa (SELECT * FROM cities WHERE country = 'CAN');

## SQL Comments

Comments are used to explain sections of SQL statements, or to prevent execution of SQL statements.


### Single line comMents

Single line comments start with --.

Any text between -- and the end of the line will be ignored (will not be executed).

The following example uses a single-line comment as an explanation:

In [ ]:
%%sql

SELECT * FROM cities LIMIT 10 -- This is a comment;

### Multi-line comments

Multi-line comments start with `/*` and end with `*/`.

Any text between `/*` and `*/` will be ignored.

The following example uses a multi-line comment as an explanation:

In [ ]:
%%sql

SELECT COUNT(name), country
FROM cities
/*
 * Adding Group by
 * Adding Order by
 */
GROUP BY country
ORDER BY COUNT(name) DESC
LIMIT 10;